bouncer/src/repo/manipulator/modeloptimizer/repo_optimizer_multipart.cpp

```diff
@@ -399,75 +917,234 @@ bool MultipartOptimizer::isTransparent(
 	return isTransparent;
 }
 
+void MultipartOptimizer::clusterMeshNodes(
+	const std::vector<repo::core::model::MeshNode>& meshes,
+	std::vector<std::vector<repo::core::model::MeshNode>>& clusters
+)
+{
+	// Takes one set of MeshNodes and groups them into N sets of MeshNodes, 
+	// where each set will form a Supermesh.
+
+	repoInfo << "Clustering " << meshes.size() << " meshes...";
+	auto start = boost::chrono::high_resolution_clock::now();
+
+	// Start by creating a metric for each mesh which will be used to group
+	// nodes by their efficiency
+
+	struct mesh {
+		repo::core::model::MeshNode node;
+		float efficiency;
+	};
+
+	std::vector<mesh> metrics(meshes.size());
+	for (size_t i = 0; i < meshes.size(); i++)
+	{
+		auto& mesh = metrics[i];
+		mesh.node = meshes[i];
+		auto bounds = mesh.node.getBoundingBox();
+		auto numVertices = mesh.node.getNumVertices();
+		auto width = bounds[1].x - bounds[0].x;
+		auto height = bounds[1].y - bounds[0].y;
+		auto length = bounds[1].z - bounds[0].z;
+		auto metric = (float)numVertices / (width + height + length);
+		mesh.efficiency = metric;
+	}
+
+	// Next order the outer nodes by their efficiency
+
+	std::sort(metrics.begin(), metrics.end(), [](mesh a, mesh b) {
+		return a.efficiency < b.efficiency;
+	});
+
+	// Bin the nodes based on vertex count. The bottom 20% will form pure-LOD
+	// groups, while the top 80% will be spatialised.
+
+	auto totalVertices = 0;
+	for (auto& mesh : meshes) {
+		totalVertices += mesh.getNumVertices();
+	}
+
+	auto modelLowerThreshold = std::max<int>(65536, (int)(totalVertices * 0.2f));
+
+	auto binVertexSize = 65536;
+	std::vector<repo::core::model::MeshNode> bin;
+	auto binVertexCount = 0;
+	auto binsVertexCount = 0;
+	
+	int i = 0;
+	for (; i < metrics.size(); )
```

Comment:
The only downside of a do..while is it will throw if the meshes count is zero. It would be pointless to call clusterMeshNodes without any meshes, but there isn't actually anything stopping that from happening!
I have re-written it to use a while loop though, which does look much more sensible without the odd break, and the increment no longer needs to be explained, as its now just used as it would be typically in such a loop.